openapi-codegen/core/src/main/scala/sttp/tapir/codegen/SchemaGenerator.scala

```diff
@@ -0,0 +1,227 @@
+  Discriminator,
+  OpenapiSchemaAllOf,
+  OpenapiSchemaAny,
+  OpenapiSchemaAnyOf,
+  OpenapiSchemaArray,
+  OpenapiSchemaConstantString,
+  OpenapiSchemaEnum,
+  OpenapiSchemaField,
+  OpenapiSchemaMap,
+  OpenapiSchemaNot,
+  OpenapiSchemaObject,
+  OpenapiSchemaOneOf,
+  OpenapiSchemaRef,
+  OpenapiSchemaSimpleType
+}
+
+import scala.collection.mutable
+
+object SchemaGenerator {
+
+  def generateSchemas(
+      doc: OpenapiDocument,
+      allSchemas: Map[String, OpenapiSchemaType],
+      fullModelPath: String,
+      jsonSerdeLib: JsonSerdeLib,
+      maxSchemasPerFile: Int
+  ): Seq[String] = {
+    def schemaContainsAny(schema: OpenapiSchemaType): Boolean = schema match {
+      case _: OpenapiSchemaAny           => true
+      case OpenapiSchemaArray(items, _)  => schemaContainsAny(items)
+      case OpenapiSchemaMap(items, _)    => schemaContainsAny(items)
+      case OpenapiSchemaObject(fs, _, _) => fs.values.map(_.`type`).exists(schemaContainsAny)
+      case OpenapiSchemaOneOf(types, _)  => types.exists(schemaContainsAny)
+      case OpenapiSchemaAllOf(types)     => types.exists(schemaContainsAny)
+      case OpenapiSchemaAnyOf(types)     => types.exists(schemaContainsAny)
+      case OpenapiSchemaNot(item)        => schemaContainsAny(item)
+      case _: OpenapiSchemaSimpleType | _: OpenapiSchemaEnum | _: OpenapiSchemaConstantString | _: OpenapiSchemaRef => false
+    }
+    val schemasWithAny = allSchemas.filter { case (_, schema) =>
+      schemaContainsAny(schema)
+    }
+    val maybeAnySchema: Option[(OpenapiSchemaType, String)] =
+      if (schemasWithAny.isEmpty) None
+      else if (jsonSerdeLib == JsonSerdeLib.Circe)
+        Some(
+          OpenapiSchemaAny(
+            false
+          ) -> "implicit lazy val anyTapirSchema: sttp.tapir.Schema[io.circe.Json] = sttp.tapir.Schema.any[io.circe.Json]"
+        )
+      else throw new NotImplementedError("any not implemented for json libs other than circe")
+    val openApiSchemasWithTapirSchemas = doc.components
+      .map(_.schemas.map {
+        case (name, _: OpenapiSchemaEnum) =>
+          name -> s"implicit lazy val ${BasicGenerator.uncapitalise(name)}TapirSchema: sttp.tapir.Schema[$name] = sttp.tapir.Schema.derived"
+        case (name, obj: OpenapiSchemaObject) => name -> schemaForObject(name, obj)
+        case (name, schema: OpenapiSchemaMap) => name -> schemaForMap(name, schema)
+        case (name, schema: OpenapiSchemaOneOf) =>
+          name -> genADTSchema(name, schema, if (fullModelPath.isEmpty) None else Some(fullModelPath))
+        case (n, x) => throw new NotImplementedError(s"Only objects, enums, maps and oneOf supported! (for $n found ${x})")
+      })
+      .toSeq
+      .flatMap(maybeAnySchema.toSeq ++ _)
+      .toMap
+
+    val groupedByRing = constructRings(allSchemas)
+    val orderedLayers = orderLayers(groupedByRing)
+    val foldedLayers = foldLayers(maxSchemasPerFile)(orderedLayers)
```

Comment:
This would've all been super-easy and could've just been done with `.grouped(maxSize)` were it not for the fact that we're trying not to split mutually-recursive schema definitions across multiple files, which is where all this comes from...